### PR TITLE
envsubst: remove explicit subst of exported vars

### LINKED ISSUE #473: There is no need to explicitly list variables in envsubst
https://github.com/getodk/central/blob/f09c50bc3cd7537545916c25b792079c072a6ae1/files/service/scripts/start-odk.sh#L7

### PR DESCRIPTION
**Pull request by alxndrsn**
Approach suggested in https://github.com/getodk/central/pull/818#discussion_r1873810731

-----

For nginx config, a new approach is implemented with mawk.

This is similar to envsubst, but more ergonomic:

* no need to explicitly list all variables
* throw error on missing variables
* do not replace nginx vars like $host, $request_uri with empty strings (in contrast to envsubst when executed without an explicit variable list)

Risks:

There are a couple of changes which may break existing deployments:

* changing client-config.json.template
* requiring all substituted variable to be defined

Closes #473

-----

#### What has been done to verify that this works as intended?

Added tests, added tests to CI, run tests.

#### Why is this the best possible solution? Were any other approaches considered?

Considered perl, bash & sed, as they are other scripting languages present in the images in which we currently use `enbsubst`.  awk or perl seem the most suitable.  I am not aware of a reason to use one over the other.

```
# jonasal/nginx-certbot:5.4.0

* perl: v5.36.0
* awk: mawk 1.3.4 20200120
* node: not found
* bash: GNU bash, version 5.2.15(1)-release (x86_64-pc-linux-gnu)
* sed: sed (GNU sed) 4.9

# node:20.17.0-slim

* perl: v5.36.0
* awk: mawk 1.3.4 20200120
* node: v20.17.0
* bash: GNU bash, version 5.2.15(1)-release (x86_64-pc-linux-gnu)
* sed: sed (GNU sed) 4.9

# ghcr.io/enketo/enketo:7.4.0

* perl: v5.36.0
* awk: mawk 1.3.4 20200120
* node: v20.17.0
* bash: GNU bash, version 5.2.15(1)-release (x86_64-pc-linux-gnu)
* sed: sed (GNU sed) 4.9
```

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

As mentioned above, there's a risk that throwing when a variable is not defined will break existing deployments.  That wouldn't be great.

#### Does this change require updates to documentation? If so, please file an issue [here](https://github.com/getodk/docs/issues/new) and include the link below.

Maybe?  If there is some genuine risk in the previous answer.

#### Before submitting this PR, please make sure you have:

- [x] branched off and targeted the `next` branch OR only changed documentation/infrastructure (`master` is stable and used in production)
- [x] verified that any code or assets from external sources are properly credited in comments or that everything is internally sourced

### CODE REVIEW

File: files/shared/envsub.awk [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/mawk -f
+
```
[the remark this answers is not in the document]
Comment (alxndrsn) [reply files/shared/envsub.awk:2]: I think copyright is implicit; no other scripts in this repo have headers.

File: nginx.dockerfile [hunk 2]
```diff
@@ -28,7 +28,9 @@ RUN apt-get update && apt-get install -y netcat-openbsd
 
 RUN mkdir -p /usr/share/odk/nginx/
 
-COPY files/nginx/setup-odk.sh /scripts/
+COPY files/nginx/setup-odk.sh \
+     files/shared/envsub.awk \
+     /scripts/
 RUN chmod +x /scripts/setup-odk.sh
```
[the remark this answers is not in the document]
Comment (alxndrsn) [reply nginx.dockerfile:34]: I think `setup-odk.sh` is the outlier here.  The `chmod` call was introduced in #676 without explanation, and there are no other `chmod` calls in this repo, despite various places where scripts are copied from git into images.  I think this is an interesting discussion, but orthogonal to the intent of this PR.

File: files/shared/envsub.awk [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  errorCount = 0;
+}
+
+{
+  while(match($0, /\$\{[A-Z_][A-Z_0-9]*\}/) > 0) {
+    k = substr($0, RSTART+2, RLENGTH-3);
+    if(k in ENVIRON) {
+      v = ENVIRON[k];
+    } else {
+      print "ERR: var not defined on line " NR ": ${" k "}" > "/dev/fd/2";
```
[the remark this answers is not in the document]
Comment (alxndrsn) [reply files/shared/envsub.awk:24]: > would accentuate our intentions better

Agreed.

> `/dev/fd/2` is not guaranteed to be stderr. `/dev/stderr` would be a safer bet

Would this prevent redirecting?

Either way, `/dev/stderr` doesn't work as I expect **with `mawk`**:

### `gawk`

```sh
$ gawk '{ print "hi0"; print "hi1" > "/dev/stderr"; print "hi2" > "/dev/fd/2"  }' <<<'' 1>stdout.txt 2>stderr.txt; sleep 1; echo "---"; cat *.txt
---
hi1
hi2
hi0
```

```sh
$ gawk '{ print "hi0"; print "hi1" > "/dev/stderr"; print "hi2" > "/dev/fd/2"  }' <<<'' 2>/dev/null
hi0
```

```sh
$ gawk '{ print "hi0"; print "hi1" > "/dev/stderr"; print "hi2" > "/dev/fd/2"  }' <<<'' >/dev/null
hi1
hi2
```

### `mawk`

```sh
$ mawk '{ print "hi0"; print "hi1" > "/dev/stderr"; print "hi2" > "/dev/fd/2"  }' <<<'' 1>stdout.txt 2>stderr.txt; sleep 1; echo "---"; cat *.txt
---
hi2
hi0
```

```sh
$ mawk '{ print "hi0"; print "hi1" > "/dev/stderr"; print "hi2" > "/dev/fd/2"  }' <<<'' 2>/dev/null
hi0
```

```sh
$ mawk '{ print "hi0"; print "hi1" > "/dev/stderr"; print "hi2" > "/dev/fd/2"  }' <<<'' >/dev/null
hi1
hi2
```

Any idea what's going on?

File: nginx.dockerfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (alxndrsn) [reply nginx.dockerfile:34]: > Thus there is a failure scenario: when someone's `~/.gitconfig` (or somesuch) says not to honor the executability

Luckily the "someone" in question is a docker image defined in this repo, and tested in CI.  I've opened a PR to remove the `chmod` call at https://github.com/getodk/central/pull/840; discussion can continue there.

File: files/shared/envsub.awk (hunk 3 above)
[the remark this answers is not in the document]
Comment (alxndrsn) [reply files/shared/envsub.awk:24]: A possible clue:

```
$ gawk --traditional '{ print "hi0"; print "hi1" > "/dev/stderr"; print "hi2" > "/dev/fd/2"  }' <<<'' 1>stdout.txt 2>stderr.txt; sleep 1; echo "---"; cat *.txt
---
hi2
hi0
```

[the remark this answers is not in the document]
Comment (alxndrsn) [reply files/shared/envsub.awk]: This appears to be an issue with _mixing_ `"/dev/stderr"` and `"/dev/fd/2"` in the same script.

Maybe a bug in `mawk`?  Although `gawk` deliberately implements the same...

### consistently `/dev/stderr`

```sh
$ mawk '{ print "hi0"; print "hi1" > "/dev/stderr"; print "hi2" > "/dev/stderr"  }' <<<'' 1>stdout.txt 2>stderr.txt; sleep 1; echo "---"; cat *.txt
---
hi1
hi2
hi0
```

### consistently `/dev/fd/2`

```sh
$ mawk '{ print "hi0"; print "hi1" > "/dev/fd/2"; print "hi2" > "/dev/fd/2"  }' <<<'' 1>stdout.txt 2>stderr.txt; sleep 1; echo "---"; cat *.txt
---
hi1
hi2
hi0
```

Probably not important to understand this now!  I'll use `/dev/stderr` consistently.

**Comment by alxndrsn** [comment]
This should be merged after the last 2024 release goes out.

**Comment by lognaturel** [comment]
Can we remove the `gettext` dependency installed in all the images?

**Comment by alxndrsn** [comment]
> Can we remove the `gettext` dependency installed in all the images?

Good catch - that seems very likely.  I've removed `gettext` installations.

File: test/nginx/nginx.test.docker-compose.yml [hunk 4]
```diff
@@ -24,7 +24,10 @@ services:
       - enketo
     environment:
       - DOMAIN=odk-nginx.example.test
-      - SENTRY_ORG_SUBDOMAIN=example
+      - OIDC_ENABLED=
```
Comment (lognaturel) [inline test/nginx/nginx.test.docker-compose.yml:27]: `OIDC_ENABLED` is already defined below.

**Review by lognaturel — APPROVED** [review]
One small change inline and one quick confirmation -- this works fine if a user's configuration file omits certain config keys, right (e.g. the OIDC ones)? I imagine a bunch of tests would fail if that weren't the case but want to double check.

**Comment by alxndrsn** [comment]
> one quick confirmation -- this works fine if a user's configuration file omits certain config keys, right (e.g. the OIDC ones)?

I think this is handled in the production `docker-compose.yml` file by the environment defaults:

https://github.com/getodk/central/blob/279c4cb68c14854efb3ca41164451cae33f4fa7d/docker-compose.yml#L49-L76

If a value is undefined by a user, *and* omitted from these lists, then substitution would fail.

I've added an extra test case to ensure that substituting empty values is supported.

**Review by lognaturel — APPROVED** [review]
